src/renderers/wgpu/pipeline.rs

```diff
@@ -0,0 +1,347 @@
+                &texture_layout,
+            ],
+            push_constant_ranges: &[],
+        });
+
+        let mask_pipeline_layout = device.create_pipeline_layout(&PipelineLayoutDescriptor {
+            label: Some("inox2d mask pipeline layout"),
+            bind_group_layouts: &[&uniform_layout, &texture_layout],
+            push_constant_ranges: &[],
+        });
+
+        // for (mode, state) in get
+
+        let mut basic_pipelines: HashMap<BlendMode, RenderPipeline> = HashMap::new();
+        for mode in BlendMode::VALUES {
+            let basic_pipeline = create_part_pipeline(
+                device,
+                Some("inox2d basic pipeline"),
+                &pipeline_layout,
+                texture_format,
+                &device.create_shader_module(include_wgsl!("shaders/basic/basic.frag.wgsl")),
+                &device.create_shader_module(include_wgsl!("shaders/basic/basic.vert.wgsl")),
+                false,
+                blend_state_for_blend_mode(mode),
+            );
+
+            basic_pipelines.insert(mode, basic_pipeline);
+        }
+
+        let mut composite_pipelines: HashMap<BlendMode, RenderPipeline> = HashMap::new();
+        for mode in BlendMode::VALUES {
+            let composite_pipeline = create_part_pipeline(
+                device,
+                Some("inox2d composite pipeline"),
+                &pipeline_layout,
+                texture_format,
+                &device.create_shader_module(include_wgsl!("shaders/basic/composite.frag.wgsl")),
+                &device.create_shader_module(include_wgsl!("shaders/basic/composite.vert.wgsl")),
+                true,
+                blend_state_for_blend_mode(mode),
+            );
+
+            composite_pipelines.insert(mode, composite_pipeline);
+        }
+
+        let mask_pipeline = create_stencil_pipeline(
+            device,
+            Some("inox2d mask pipeline"),
+            &mask_pipeline_layout,
+            texture_format,
+            &device.create_shader_module(include_wgsl!("shaders/basic/mask.frag.wgsl")),
+            &device.create_shader_module(include_wgsl!("shaders/basic/mask.vert.wgsl")),
+        );
+
+        let min_uniform_buffer_offset_alignment =
+            device.limits().min_uniform_buffer_offset_alignment;
+
+        InoxPipeline {
+            basic_pipelines,
+            composite_pipelines,
+            mask_pipeline,
+
+            uniform_layout,
+            texture_layout,
+            texture_format,
+            uniform_alignment_needed: Uniform::min_size()
+                .get()
+                .max(min_uniform_buffer_offset_alignment.into()),
+        }
+    }
+}
+
+#[derive(ShaderType, Debug, Clone, Copy, PartialEq)]
+pub struct CameraData {
+    pub mvp: Mat4,
+}
+
+#[derive(ShaderType, Debug, Clone, Copy, PartialEq)]
+pub struct Uniform {
+    pub opacity: f32,
+    pub mult_color: Vec3,
+    pub screen_color: Vec3,
+    pub emission_strength: f32,
+    pub offset: Vec2,
+}
```

Comment:
Aren't there better names for these structs? idk, something in common that tells us these structs are meant to be passed to the WGSL shaders as uniforms.

Reply:
Imma keep the names as-is for now, maybe I'll find a better name for them later.